src/apps/lwaftr/fragmentv6_hardened.lua

```diff
@@ -0,0 +1,299 @@
+   reassembly_buf.running_length = reassembly_buf.running_length + frag_size
+
+   local restatus = reassembly_status(reassembly_buf)
+   if restatus == REASSEMBLY_OK then
+      local reassembled_packet = packet.clone(reassembly_buf.reassembly_packet)
+      free_reassembly_buf_and_pkt(fragment, frags_table)
+      return REASSEMBLY_OK, reassembled_packet
+   else
+      packet.free(fragment)
+      return restatus
+   end
+end
+
+local function packet_to_reassembly_buffer(pkt)
+   local reassembly_buf = scratch_rbuf
+   ffi.C.memset(reassembly_buf, 0, ffi.sizeof(ipv6_reassembly_buffer_t))
+   reassembly_buf.fragment_id = get_frag_id(pkt)
+   reassembly_buf.reassembly_base = ehs + ipv6_fixed_header_size
+
+   local tmplen = pkt.length
+   pkt.length = ehs + ipv6_fixed_header_size
+   local repkt = reassembly_buf.reassembly_packet
+   packet.clone_to_memory(repkt, pkt)
+   reassembly_buf.running_length = pkt.length
+
+   pkt.length = tmplen
+   --Take the next header information from the fragment
+   local next_header_base_offset = ehs + o_ipv6_next_header
+   local next_header_frag_offset = ehs + ipv6_fixed_header_size -- +0
+   repkt.data[next_header_base_offset] = pkt.data[next_header_frag_offset]
+
+   return reassembly_buf
+end
+
+-- The key is 288 bytes: source IPv6 address, destination IPv6 address, and
+-- the identification field from the IPv6 fragmentation header.
+local function hash_ipv6(key)
+   local hash = 0
+   local to_hash = ffi.cast(uint32_ptr_t, key)
+   for i=0,8 do
+      local current = to_hash[i]
+      hash = hash_32(bxor(hash, hash_32(current)))
+   end
+   return hash
+end
+
+function initialize_frag_table(max_fragmented_packets, max_pkt_frag)
+   -- Initialize module-scoped variables
+   max_frags_per_packet = max_pkt_frag
+   ipv6_reassembly_buffer_t = ffi.typeof([[
+   struct {
+       uint16_t fragment_starts[$];
+       uint16_t fragment_ends[$];
+       uint16_t fragment_count;
+       uint16_t final_start;
+       uint16_t reassembly_base;
+       uint32_t fragment_id;
+       uint32_t running_length;
+       struct packet reassembly_packet;
+   } __attribute((packed))]],
+   max_frags_per_packet, max_frags_per_packet)
+   scratch_rbuf = ipv6_reassembly_buffer_t()
+
+   local max_occupy = 0.9
+   local params = {
+      key_type = ffi.typeof(ipv6_fragment_key_t),
+      value_type = ffi.typeof(ipv6_reassembly_buffer_t),
+      hash_fn = hash_ipv6,
+      initial_size = math.ceil(max_fragmented_packets / max_occupy),
+      max_occupancy_rate = max_occupy
+   }
+   return ctable.new(params)
+end
+
+function cache_fragment(frags_table, fragment)
+   local key = get_key(fragment)
+   local ptr = frags_table:lookup_ptr(key)
+   if not ptr then
+      local reassembly_buf = packet_to_reassembly_buffer(fragment)
+      frags_table:add_with_random_ejection(key, reassembly_buf, false)
+      ptr = frags_table:lookup_ptr(key)
+   end
+   local s,p = attempt_reassembly(frags_table, ptr.value, fragment)
+   return s, p
```

Comment:
Since `s` and `p` are not used, can return the result of `attempt_reassembly`.

```lua
return attempt_reassembly(frags_table, ptr.value, fragment)
```